R/preprocessing.R

```diff
@@ -1526,3 +1526,165 @@ RegressOutNB <- function(
   object@scale.data <- pr
   return(object)
 }
+
+
+
+#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+# Proposed feature: subset by barcode distribution inflection points
+#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+#' Subsets a Seurat Object By Distribution of UMIs
+#'
+#' This function calculates an adaptive inflection point ("knee") of the barcode distribution
+#' for each sample group (as identified by `object@meta.data$orig.ident`).
+#' Returns the subsetted `object` with cells that passed the calculated threshold.
+#'
+#' The assumed assay is "RNA", and thus uses the "nCount_RNA" distribution for
+#' calculating the appropriate cutoff and returning a subsetted object.
+#'
+#' @param object Seurat object
+#' @param umi.column Column to use as proxy for barcodes ("nCount_RNA" by default)
+#' @param show.plot Whether to produce diagnostic plot showing inflection point and barcode distribution
+#' @param threshold.low Ignore barcodes of rank below this threshold in inflection calculation
+#' @param threshold.high Ignore barcodes of rank above this threshold in inflection calculation
+#' @param return.plot Returns the ggplot2 object in lieu of the subsetted Seurat object (useful for visually testing thresholds)
+#' @param return.cells Returns the vector of cells which pass the defined barcode thresholds
+#' @param return.inflections Returns a tibble with the grouping variable, data list-column with cell, umi.column, and rank, the calculated inflection point and associated umi_threshold, and a list-column of the cells which passed the calculated threshold per group
+#'
+#' @return Return Seurat object subsetted. Appends to the workflow slot 'nCount_RNA_threshold'
+#' with a `data.frame` showing the calculated thresholds per sample group that determine the passing cells.
+#'
+#' @importFrom dplyr mutate arrange group_by summarise select inner_join filter
+#' @importFrom tidyr nest
+#' @importFrom purrr map_int map2_dbl map2
+#' @importFrom ggplot2 ggplot geom_line theme_bw geom_vline
+#' @importFrom rlang sym
+#' @import tibble
+#'
+#' @rdname Preprocessing
+#'
+#' @export
+#' 
+#' @examples
+#' \dontrun{
+#' SubsetByBarcodeDistribution(pbmc_small)
+#' }
+
+SubsetByBarcodeDistribution <- function(object,
+                                       umi.column = 'nCount_RNA',
+                                       threshold.low = NULL,
+                                       threshold.high = NULL,
+                                       show.plot = FALSE,                                       
+                                       return.plot = FALSE,
+                                       return.cells = FALSE,
+                                       return.inflections = FALSE) {
+    ## Make sure there is a column for "cell" to subset the object by
+    if (!("cell" %in% colnames(object@meta.data))) {
+        object@meta.data$cell <- rownames(object@meta.data)
```

Comment:
Please use the `[[` and `[[<-` operators for getting and setting meta data

Reply:
Could I get an example usage? I'm not familiar with this..thanks!

Reply:
```R
colnames(x = pbmc_small[[]])
pbmc_small[['nCount_RNA']]
pbmc_small[['groups']] <- sample(x = c('g1', 'g2'), size = ncol(x = pbmc_small), replace = TRUE)
pbmc_small[['groups', drop = TRUE]]
```